README.md
---

# S5plus Lib Javascript SDK

An S5plus Lib Javascript for Decentralized content-addressed storage network S5.

## Installing

1. Clone this repo.
2. cd s5plus-lib
3. yarn install



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

For the latest beta changes, see [CHANGELOG-BETA.md](./CHANGELOG-BETA.md).

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes:

- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Removed` for now removed features.

## Unreleased

## [0.1.0-beta1] 27.11.2023

### Added




src/utils/seed.ts
import { Buffer } from "buffer";
import * as sodium from 'libsodium-wrappers';
import { blake3 } from "@noble/hashes/blake3";
import { wordlist } from './wordlist.js';

//import { encodeBase64URL } from "./tools.js";
//import { CryptoImplementation, KeyPairEd25519, encodeBase64URL } from 's5-crypto-utils';
import { CryptoImplementation, KeyPairEd25519 } from 's5-crypto-utils';

export const SEED_LENGTH = 16;
export const SEED_WORDS_LENGTH = 13;
export const CHECKSUM_WORDS_LENGTH = 2;
export const PHRASE_LENGTH = SEED_WORDS_LENGTH + CHECKSUM_WORDS_LENGTH;

/**
 * Encodes a Uint8Array into a Base64URL string.
 * @param input - The Uint8Array to be encoded.
 * @returns The Base64URL-encoded string.
 */
export function encodeBase64URL(input: Uint8Array): string {
  const base64 = btoa(String.fromCharCode(...input));

  return base64.replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");
}

/**
 * Generate a seed from a mnemonic phrase.
 * @param {string} phrase - The mnemonic phrase.
 * @returns {Promise<Uint8Array>} - The generated seed as a Uint8Array.
 */
export async function generateSeedFromPhrase(phrase: string): Promise<Uint8Array> {
  const sanitizedPhrase = sanitizePhrase(phrase);
  const validatedPhrase = await validatePhrase(sanitizedPhrase);

  const hasher = await blake3.create({dkLen: 32});
  hasher.update(validatedPhrase);
  const b3Seed = hasher.digest(); 

  return b3Seed;
}

/**
 * Generates a random passphrase.
 * @returns A promise that resolves to the generated passphrase.
 */
export async function generatePhrase(): Promise<string> {
  await sodium.ready;
  const seedWords = new Uint16Array(SEED_WORDS_LENGTH);

  // Populate the seed words from the random values.
  for (let i = 0; i < SEED_WORDS_LENGTH; i++) {
    seedWords[i] = sodium.randombytes_uniform(wordlist.length);

    let numBits = 10;
    // For the 1st word, only the first 256 words are considered valid.
    if (i === 0) {
      numBits = 8;
    }
    seedWords[i] = seedWords[i] % (1 << numBits);
  }

  // Generate checksum from hash of the seed.
  const checksumWords = generateChecksumWordsFromSeedWords(
    seedWords
  );

  const phraseWords = new Array<string>(PHRASE_LENGTH);

  for (let i = 0; i < SEED_WORDS_LENGTH; i++) {
    phraseWords[i] = wordlist[seedWords[i]];
  }
  for (let i = 0; i < CHECKSUM_WORDS_LENGTH; i++) {
    phraseWords[i + SEED_WORDS_LENGTH] = wordlist[checksumWords[i]];
  }

  return phraseWords.join(" ");
}

/**
 * Sanitize the input phrase by trimming and converting to lowercase.
 * @param phrase - The input passphrase.
 * @returns The sanitized phrase.
 */
export function sanitizePhrase(phrase: string): string {
  return phrase.trim().toLowerCase();
}

/**
 * Validate and convert a passphrase into a Uint8Array seed.
 * @param phrase - The input passphrase.
 * @throws Error if the phrase is invalid.
 * @returns The Uint8Array seed.
 */
export async function validatePhrase(phrase: string): Promise<Uint8Array> {
  await sodium.ready;

  phrase = sanitizePhrase(phrase);
  const phraseWords = phrase.split(' ');

  if (phraseWords.length !== PHRASE_LENGTH) {
    throw new Error(`Phrase must be ${PHRASE_LENGTH} words long`);
  }

  // Build the seed from words.
  const seedWords = new Uint16Array(SEED_WORDS_LENGTH);

  let i = 0;
  for (const word of phraseWords) {
    // Check word length.
    if (word.length < 3) {
      throw new Error(`Word ${i + 1} is not at least 3 letters long`);
    }

    // Check word prefix.
    const prefix = word.substring(0, 3);
    let bound = wordlist.length;
    if (i === 0) {
      bound = 256;
    }
    let found = -1;
    for (let j = 0; j < bound; j++) {
      const curPrefix = wordlist[j].substring(0, 3);
      if (curPrefix === prefix) {
        found = j;
        break;
      }
    }
    if (found < 0) {
      if (i === 0) {
        throw new Error(`Prefix for word ${i + 1} must be found in the first 256 words of the wordlist`);
      } else {
        throw new Error(`Unrecognized prefix "${prefix}" at word ${i + 1}, not found in wordlist`);
      }
    }

    seedWords[i] = found;

    i++;
    if (i >= SEED_WORDS_LENGTH) break;
  }

  // Validate checksum.
  const checksumWords = generateChecksumWordsFromSeedWords(seedWords);
  for (let i = 0; i < CHECKSUM_WORDS_LENGTH; i++) {
    const prefix = wordlist[checksumWords[i]].substring(0, 3);
    if (phraseWords[i + SEED_WORDS_LENGTH].substring(0, 3) !== prefix) {
      throw new Error(`Word "${phraseWords[i + SEED_WORDS_LENGTH + 1]}" is not a valid checksum for the seed`);
    }
  }

  return seedWordsToSeed(seedWords);
}

/**
 * Generate checksum words from seed words.
 * @param seedWords - The seed words as Uint16Array.
 * @throws Error if the input seed is not of the expected length.
 * @returns The checksum words as Uint16Array.
 */
export function generateChecksumWordsFromSeedWords(seedWords: Uint16Array): Uint16Array {
  if (seedWords.length !== SEED_WORDS_LENGTH) {
    throw new Error(`Input seed was not of length ${SEED_WORDS_LENGTH}`);
  }

  const seed = seedWordsToSeed(seedWords);
  const h = blake3(seed, { dkLen: SEED_LENGTH});
  const checksumWords = hashToChecksumWords(h);

  return checksumWords;
}

/**
 * Convert a hash to checksum words.
 * @param h - The input hash as Uint8Array.
 * @returns The checksum words as Uint16Array.
 */
export function hashToChecksumWords(h: Uint8Array): Uint16Array {
  let word1 = (h[0] << 8);
  word1 += h[1];
  word1 >>= 6;
  let word2 = (h[1] << 10);
  word2 &= 0xffff;
  word2 += (h[2] << 2);
  word2 >>= 6;

  return new Uint16Array([word1, word2]);
}

/**
 * Convert seed words to a seed as Uint8Array.
 * @param seedWords - The seed words as Uint16Array.
 * @throws Error if the input seed words are not of the expected length.
 * @returns The seed as Uint8Array.
 */
export function seedWordsToSeed(seedWords: Uint16Array): Uint8Array {
  if (seedWords.length !== SEED_WORDS_LENGTH) {
    throw new Error(`Input seed was not of length ${SEED_WORDS_LENGTH}`);
  }

  // We are getting 16 bytes of entropy.
  const bytes = new Uint8Array(SEED_LENGTH);
  let curByte = 0;
  let curBit = 0;

  for (let i = 0; i < SEED_WORDS_LENGTH; i++) {
    const word = seedWords[i];
    let wordBits = 10;
    if (i === 0) {
      wordBits = 8;
    }

    // Iterate over the bits of the 10- or 8-bit word.
    for (let j = 0; j < wordBits; j++) {
      const bitSet = (word & (1 << (wordBits - j - 1))) > 0;

      if (bitSet) {
        bytes[curByte] |= (1 << (8 - curBit - 1));
      }

      curBit += 1;
      if (curBit >= 8) {
        curByte += 1;
        curBit = 0;
      }
    }
  }

  return bytes;
}

/**
 * Define the KeyPairAndSeed type.
 */
export interface KeyPairAndSeed {
  privateKey: string;
  publicKey: string;
  publicKeyRaw: string;
  seed: string;
}

/**
 * Generates a key pair and seed for cryptographic purposes.
 * @param length - The length of the seed (default: 32).
 * @returns An object containing the generated key pair and seed.
 */
export async function genKeyPairAndSeed(length = 32): Promise<KeyPairAndSeed> {
  await sodium.ready; // Make sure sodium is ready for use

  const crypto = new CryptoImplementation(); 
  const randomSeed =  await crypto.generateRandomBytes(length);
  const keyPair = await crypto.newKeyPairEd25519({ seed: randomSeed });

  return {
    privateKey: encodeBase64URL(await keyPair.extractBytes()),
    publicKey: encodeBase64URL(await keyPair.publicKey()),
    publicKeyRaw: encodeBase64URL((await keyPair.publicKey()).slice(1)),
    seed: Buffer.from(randomSeed).toString('hex'),
  };
}

/**
 * Define the KeyPair interface.
 */
export interface KeyPair {
  privateKey: string;
  publicKey: string;
  publicKeyRaw: string;
}

/**
 * Generates a key pair from a given seed using libsodium-wrappers.
 * @param seedString - The seedString as a string.
 * @returns A Promise that resolves to an object containing the generated public and private keys as hexadecimal strings.
 */
export async function genKeyPairFromSeed(seedString: string): Promise<KeyPair> {
  await sodium.ready;

  const seed = await generateSeedFromPhrase(seedString);

  const crypto = new CryptoImplementation(); 
  const keyPair = await crypto.newKeyPairEd25519({ seed: seed });

  return {
    privateKey: encodeBase64URL(await keyPair.extractBytes()),
    publicKey: encodeBase64URL(await keyPair.publicKey()),
    publicKeyRaw: encodeBase64URL((await keyPair.publicKey()).slice(1))
  }; 
}

/**
 * Derives an Ed25519 key pair from a master key and a data seed.
 * @param masterKey - The master key as a Uint8Array or string.
 * @param dataSeed - The data seed as a Uint8Array or string.
 * @param deriveLength - (Optional) The length of the derived key in bytes.
 * @returns A Promise that resolves to an object containing the derived Ed25519 key pair with publicKey and privateKey.
 */
export async function deriveEd25519(masterKey: Uint8Array | string, dataSeed: Uint8Array | string, deriveLength?: number): Promise<{ privateKey: string, publicKey: string }> {
  await sodium.ready;
  let hasher;
  if (deriveLength) {
    hasher = blake3.create({ dkLen: deriveLength });
  } else {
    hasher = blake3.create({});
  }

  const masterKeyBytes = typeof masterKey === "string" ? new TextEncoder().encode(masterKey) : masterKey;
  const dataSeedBytes = typeof dataSeed === "string" ? new TextEncoder().encode(dataSeed) : dataSeed;
  hasher.update(new Uint8Array([...masterKeyBytes, ...dataSeedBytes]));
  const saltEd25519Key = hasher.digest();
  const keyPair = sodium.crypto_sign_seed_keypair(saltEd25519Key);
  const kp = new KeyPairEd25519(keyPair.privateKey);

  return {
    privateKey: encodeBase64URL(await kp.extractBytes()),
    publicKey: encodeBase64URL(await kp.publicKey()),
  };
}



src/utils/wordlist.ts
/// Wordlist where each word has a unique 3-letter prefix. There are 1024 words
/// so that each word gives us 10 bits of entropy, which lets us get 16 bytes of
/// entropy from 13 words. A small wordlist also helps us keep our memory
/// footprint lower.

export const uniquePrefixLen = 3;

export const wordlist = [
  'abandon',
  'ability',
  'able',
  'about',
  'abroad',
  'absence',
  'abuse',
  'academy',
  'accept',
  'achieve',
  'acid',
  'acquire',
  'across',
  'action',
  'adapt',
  'address',
  'adjust',
  'admit',
  'adopt',
  'adult',
  'advice',
  'affect',
  'afraid',
  'after',
  'against',
  'agency',
  'ago',
  'agree',
  'ahead',
  'airport',
  'aisle',
  'alarm',
  'album',
  'alcohol',
  'alive',
  'allow',
  'almost',
  'along',
  'already',
  'also',
  'alter',
  'always',
  'amazing',
  'amid',
  'among',
  'analyst',
  'ancient',
  'angry',
  'animal',
  'ankle',
  'annual',
  'another',
  'answer',
  'anxiety',
  'anyone',
  'apart',
  'apology',
  'appear',
  'area',
  'argue',
  'arise',
  'army',
  'around',
  'arrive',
  'article',
  'aside',
  'asleep',
  'aspect',
  'assume',
  'atop',
  'attack',
  'auction',
  'audio',
  'aunt',
  'author',
  'average',
  'avoid',
  'away',
  'awful',
  'baby',
  'back',
  'badly',
  'bake',
  'ball',
  'bank',
  'barely',
  'base',
  'battle',
  'beat',
  'because',
  'bedroom',
  'beer',
  'before',
  'begin',
  'behind',
  'being',
  'believe',
  'benefit',
  'best',
  'between',
  'beyond',
  'bias',
  'bike',
  'billion',
  'bind',
  'biology',
  'bird',
  'bishop',
  'bite',
  'bizarre',
  'black',
  'blend',
  'blind',
  'blood',
  'blue',
  'board',
  'body',
  'boil',
  'bold',
  'bomb',
  'bone',
  'book',
  'born',
  'boss',
  'both',
  'bounce',
  'bowl',
  'box',
  'boy',
  'brain',
  'break',
  'bring',
  'brother',
  'brush',
  'bubble',
  'buck',
  'budget',
  'bug',
  'build',
  'bullet',
  'bunch',
  'burn',
  'busy',
  'butter',
  'buyer',
  'cable',
  'cafe',
  'cage',
  'cake',
  'call',
  'camera',
  'can',
  'capital',
  'care',
  'case',
  'catch',
  'cause',
  'cave',
  'cease',
  'ceiling',
  'cell',
  'center',
  'certain',
  'change',
  'check',
  'child',
  'choose',
  'chronic',
  'church',
  'circle',
  'city',
  'civil',
  'class',
  'clear',
  'client',
  'close',
  'club',
  'coach',
  'cocaine',
  'code',
  'coffee',
  'coin',
  'college',
  'come',
  'control',
  'cook',
  'copy',
  'corner',
  'cost',
  'cotton',
  'could',
  'cover',
  'cow',
  'crazy',
  'create',
  'crime',
  'crowd',
  'crucial',
  'crystal',
  'cue',
  'culture',
  'cup',
  'current',
  'custom',
  'cute',
  'cycle',
  'dad',
  'daily',
  'damage',
  'danger',
  'dark',
  'data',
  'dawn',
  'day',
  'death',
  'debate',
  'decide',
  'deep',
  'defense',
  'degree',
  'deliver',
  'demand',
  'deny',
  'depend',
  'despite',
  'detail',
  'develop',
  'diamond',
  'dictate',
  'diet',
  'differ',
  'digital',
  'dilemma',
  'dinner',
  'dip',
  'direct',
  'disease',
  'divide',
  'doctor',
  'dog',
  'dollar',
  'domain',
  'donor',
  'door',
  'dose',
  'doubt',
  'down',
  'dozen',
  'draw',
  'dream',
  'drive',
  'drop',
  'drug',
  'dry',
  'duck',
  'dump',
  'during',
  'dust',
  'duty',
  'dying',
  'dynamic',
  'each',
  'eager',
  'early',
  'easy',
  'eating',
  'echo',
  'economy',
  'edge',
  'editor',
  'educate',
  'effect',
  'egg',
  'ego',
  'either',
  'elbow',
  'elderly',
  'element',
  'elite',
  'else',
  'embrace',
  'emerge',
  'emotion',
  'empty',
  'enable',
  'endure',
  'energy',
  'enforce',
  'engage',
  'enhance',
  'enjoy',
  'enough',
  'ensure',
  'enter',
  'episode',
  'equal',
  'error',
  'escape',
  'essay',
  'estate',
  'ethnic',
  'even',
  'evident',
  'evolve',
  'exactly',
  'except',
  'execute',
  'exhibit',
  'exist',
  'exotic',
  'expect',
  'extend',
  'eye',
  'fabric',
  'fact',
  'fade',
  'fail',
  'fall',
  'family',
  'fantasy',
  'farm',
  'fast',
  'father',
  'fault',
  'favor',
  'fear',
  'federal',
  'feel',
  'fellow',
  'female',
  'fence',
  'fever',
  'fewer',
  'fiber',
  'fiction',
  'field',
  'figure',
  'film',
  'find',
  'first',
  'fish',
  'fitness',
  'fixed',
  'flag',
  'flee',
  'flight',
  'floor',
  'fluid',
  'flying',
  'focus',
  'fog',
  'follow',
  'food',
  'force',
  'foster',
  'frame',
  'free',
  'friend',
  'from',
  'fruit',
  'fuel',
  'full',
  'fund',
  'fury',
  'future',
  'gain',
  'gallery',
  'game',
  'gang',
  'garden',
  'gather',
  'gay',
  'gaze',
  'gear',
  'general',
  'gesture',
  'ghost',
  'giant',
  'gift',
  'girl',
  'give',
  'glass',
  'glimpse',
  'global',
  'goal',
  'gold',
  'good',
  'gospel',
  'govern',
  'grade',
  'great',
  'grip',
  'group',
  'guard',
  'guess',
  'guide',
  'gulf',
  'gut',
  'guy',
  'gym',
  'habit',
  'hair',
  'half',
  'hammer',
  'hand',
  'happen',
  'hard',
  'hate',
  'haul',
  'have',
  'hazard',
  'hear',
  'heel',
  'height',
  'help',
  'hence',
  'here',
  'hide',
  'high',
  'hill',
  'himself',
  'hint',
  'hip',
  'hire',
  'history',
  'hockey',
  'hold',
  'home',
  'honor',
  'hook',
  'hope',
  'horse',
  'host',
  'hotel',
  'house',
  'however',
  'huge',
  'human',
  'hundred',
  'hurt',
  'husband',
  'ice',
  'icon',
  'idea',
  'ignore',
  'image',
  'immune',
  'impact',
  'include',
  'indeed',
  'inform',
  'inherit',
  'initial',
  'injury',
  'inmate',
  'inner',
  'input',
  'inquiry',
  'inside',
  'into',
  'involve',
  'iron',
  'island',
  'isolate',
  'issue',
  'item',
  'itself',
  'jacket',
  'jail',
  'jaw',
  'jazz',
  'jeans',
  'jewelry',
  'job',
  'join',
  'joke',
  'journal',
  'joy',
  'judge',
  'juice',
  'jump',
  'junior',
  'jury',
  'just',
  'keep',
  'key',
  'kick',
  'kill',
  'kind',
  'kiss',
  'kitchen',
  'knee',
  'knife',
  'know',
  'labor',
  'lack',
  'lady',
  'lake',
  'lamp',
  'land',
  'large',
  'last',
  'later',
  'laugh',
  'layer',
  'leave',
  'lecture',
  'left',
  'legal',
  'lemon',
  'length',
  'less',
  'letter',
  'level',
  'library',
  'license',
  'life',
  'light',
  'like',
  'limit',
  'line',
  'lion',
  'liquid',
  'listen',
  'little',
  'live',
  'loan',
  'lobby',
  'local',
  'logic',
  'long',
  'look',
  'lord',
  'lose',
  'lots',
  'loud',
  'love',
  'lower',
  'loyalty',
  'lucky',
  'lunch',
  'machine',
  'madness',
  'magic',
  'main',
  'major',
  'make',
  'male',
  'mama',
  'many',
  'market',
  'matter',
  'maximum',
  'maybe',
  'mean',
  'medical',
  'meet',
  'melt',
  'member',
  'mention',
  'merely',
  'message',
  'method',
  'middle',
  'might',
  'million',
  'minute',
  'mirror',
  'miss',
  'mixture',
  'mobile',
  'model',
  'moment',
  'money',
  'moon',
  'more',
  'most',
  'mother',
  'mouth',
  'move',
  'much',
  'mud',
  'murder',
  'must',
  'mutual',
  'myself',
  'myth',
  'nail',
  'naked',
  'name',
  'narrow',
  'nasty',
  'nation',
  'navy',
  'nearly',
  'neck',
  'need',
  'neither',
  'nervous',
  'nest',
  'network',
  'neutral',
  'never',
  'news',
  'next',
  'nice',
  'night',
  'nobody',
  'noise',
  'nominee',
  'none',
  'noon',
  'north',
  'nose',
  'nothing',
  'novel',
  'nowhere',
  'nuclear',
  'number',
  'nurse',
  'nut',
  'oak',
  'object',
  'observe',
  'obtain',
  'obvious',
  'occur',
  'ocean',
  'odds',
  'office',
  'often',
  'okay',
  'old',
  'once',
  'ongoing',
  'onion',
  'only',
  'onto',
  'open',
  'opinion',
  'oppose',
  'option',
  'orange',
  'orbit',
  'order',
  'organic',
  'origin',
  'other',
  'ought',
  'ounce',
  'ours',
  'outside',
  'over',
  'owner',
  'oxygen',
  'package',
  'page',
  'pain',
  'pale',
  'panel',
  'paper',
  'part',
  'pass',
  'patient',
  'pause',
  'payment',
  'peace',
  'peer',
  'penalty',
  'people',
  'pepper',
  'percent',
  'pet',
  'phase',
  'photo',
  'phrase',
  'physics',
  'piano',
  'pick',
  'piece',
  'pilot',
  'pink',
  'pioneer',
  'pipe',
  'pitch',
  'pizza',
  'place',
  'please',
  'plot',
  'plus',
  'pocket',
  'poem',
  'point',
  'poke',
  'policy',
  'pond',
  'poor',
  'popular',
  'portion',
  'pose',
  'potato',
  'pound',
  'poverty',
  'power',
  'prayer',
  'prepare',
  'price',
  'problem',
  'public',
  'pull',
  'pump',
  'punish',
  'purpose',
  'push',
  'put',
  'puzzle',
  'quality',
  'queen',
  'quite',
  'quote',
  'rabbit',
  'race',
  'radio',
  'rage',
  'raise',
  'rally',
  'range',
  'rapidly',
  'rare',
  'rate',
  'really',
  'rebuild',
  'receive',
  'reduce',
  'reflect',
  'region',
  'reject',
  'relate',
  'remain',
  'rent',
  'report',
  'require',
  'result',
  'return',
  'reveal',
  'reward',
  'rhythm',
  'ribbon',
  'rich',
  'ride',
  'rifle',
  'right',
  'ring',
  'riot',
  'risk',
  'ritual',
  'river',
  'road',
  'robot',
  'rock',
  'role',
  'romance',
  'room',
  'rope',
  'route',
  'royal',
  'rubber',
  'ruin',
  'rule',
  'rumor',
  'running',
  'rural',
  'rush',
  'sacred',
  'safety',
  'sail',
  'salt',
  'same',
  'sand',
  'satisfy',
  'sauce',
  'save',
  'say',
  'scale',
  'scene',
  'school',
  'science',
  'score',
  'screen',
  'season',
  'second',
  'see',
  'segment',
  'seize',
  'sell',
  'seminar',
  'send',
  'service',
  'session',
  'settle',
  'several',
  'share',
  'sheet',
  'ship',
  'should',
  'shrug',
  'shut',
  'shy',
  'sibling',
  'sick',
  'side',
  'sign',
  'silence',
  'simply',
  'since',
  'sister',
  'site',
  'size',
  'skill',
  'skull',
  'sky',
  'slave',
  'sleep',
  'slip',
  'slowly',
  'small',
  'smell',
  'smile',
  'smoke',
  'snap',
  'sneak',
  'snow',
  'soap',
  'social',
  'sodium',
  'soft',
  'soil',
  'soldier',
  'some',
  'song',
  'soon',
  'sort',
  'source',
  'space',
  'speak',
  'sphere',
  'spirit',
  'split',
  'sport',
  'spring',
  'spy',
  'square',
  'state',
  'step',
  'still',
  'story',
  'street',
  'student',
  'style',
  'subject',
  'such',
  'sudden',
  'suffer',
  'suggest',
  'suit',
  'summer',
  'sunset',
  'support',
  'sure',
  'suspect',
  'swallow',
  'sweet',
  'swing',
  'sword',
  'symbol',
  'system',
  'table',
  'tactic',
  'tail',
  'take',
  'talk',
  'tank',
  'tape',
  'target',
  'task',
  'tax',
  'team',
  'tech',
  'teen',
  'tell',
  'temple',
  'tend',
  'term',
  'test',
  'text',
  'that',
  'they',
  'this',
  'those',
  'through',
  'thus',
  'ticket',
  'tide',
  'tiger',
  'till',
  'time',
  'tiny',
  'tired',
  'tissue',
  'title',
  'tobacco',
  'today',
  'toilet',
  'toll',
  'tomato',
  'tonight',
  'tool',
  'topic',
  'toss',
  'total',
  'tough',
  'toward',
  'toxic',
  'toy',
  'trade',
  'tree',
  'trial',
  'trouble',
  'true',
  'try',
  'tube',
  'tuck',
  'tumor',
  'tunnel',
  'turn',
  'twenty',
  'twice',
  'type',
  'ugly',
  'unable',
  'uncle',
  'under',
  'unfair',
  'unhappy',
  'union',
  'unknown',
  'unless',
  'until',
  'unusual',
  'update',
  'upon',
  'upper',
  'upset',
  'urban',
  'urge',
  'used',
  'usually',
  'utility',
  'vaccine',
  'value',
  'vanish',
  'various',
  'vast',
  'vehicle',
  'venture',
  'very',
  'vessel',
  'veteran',
  'victim',
  'video',
  'view',
  'village',
  'violent',
  'virus',
  'visit',
  'vital',
  'voice',
  'volume',
  'vote',
  'wage',
  'wait',
  'wake',
  'walk',
  'want',
  'warm',
  'waste',
  'water',
  'wave',
  'way',
  'wear',
  'web',
  'wedding',
  'week',
  'weight',
  'well',
  'west',
  'what',
  'when',
  'which',
  'whole',
  'why',
  'wide',
  'wife',
  'will',
  'window',
  'wipe',
  'wire',
  'wish',
  'with',
  'wolf',
  'woman',
  'wonder',
  'wood',
  'work',
  'would',
  'wrap',
  'write',
  'wrong',
  'yard',
  'year',
  'yellow',
  'yield',
  'you',
  'zone',
];
